File: flutter_client/lib/models/message.dart
// ABOUTME: Chat message model for displaying conversation history.
// ABOUTME: Supports user messages, assistant messages, tool calls, and errors.

import 'geo_feature.dart';

enum MessageKind {
  user,
  assistant,
  toolCall,
  toolResult,
  subagentCall,
  subagentResult,
  geoFeature,
  error,
}

abstract class MessageContent {
  const MessageContent();
}

class TextContent extends MessageContent {
  final String text;
  const TextContent(this.text);
}

class ToolCallContent extends MessageContent {
  final String toolName;
  final Map<String, dynamic> input;
  const ToolCallContent({
    required this.toolName,
    required this.input,
  });
}

class ToolResultContent extends MessageContent {
  final String toolName;
  final Map<String, dynamic>? result;
  final String? error;
  const ToolResultContent({
    required this.toolName,
    this.result,
    this.error,
  });
}

class SubagentCallContent extends MessageContent {
  final String agentName;
  const SubagentCallContent({
    required this.agentName,
  });
}

class SubagentResultContent extends MessageContent {
  final String agentName;
  final String? error;
  const SubagentResultContent({
    required this.agentName,
    this.error,
  });
}

class GeoFeatureContent extends MessageContent {
  final GeoFeature feature;
  const GeoFeatureContent(this.feature);
}

class Message {
  final String id;
  final MessageKind kind;
  final MessageContent content;
  final DateTime timestamp;

  Message({
    required this.id,
    required this.kind,
    required this.content,
    required this.timestamp,
  });
}

File: react_client/src/models/message.ts
// ABOUTME: Chat message model for displaying conversation history.
// ABOUTME: Uses discriminated unions for type-safe content handling.

import type { GeoFeature } from "./geo_feature";

export type MessageKind =
  | "user"
  | "assistant"
  | "toolCall"
  | "toolResult"
  | "subagentCall"
  | "subagentText"
  | "subagentResult"
  | "geoFeature"
  | "error";

export type MessageContent =
  | { type: "text"; text: string }
  | { type: "toolCall"; toolName: string; input: Record<string, unknown> }
  | {
      type: "toolResult";
      toolName: string;
      result?: Record<string, unknown>;
      error?: string;
    }
  | { type: "subagentCall"; agentName: string }
  | { type: "subagentText"; agentName: string; text: string }
  | { type: "subagentResult"; agentName: string; error?: string }
  | { type: "geoFeature"; feature: GeoFeature };

export interface Message {
  id: string;
  kind: MessageKind;
  content: MessageContent;
  timestamp: Date;
}

export function createTextMessage(
  id: string,
  kind: MessageKind,
  text: string
): Message {
  return {
    id,
    kind,
    content: { type: "text", text },
    timestamp: new Date(),
  };
}

export function createToolCallMessage(
  id: string,
  toolName: string,
  input: Record<string, unknown>
): Message {
  return {
    id,
    kind: "toolCall",
    content: { type: "toolCall", toolName, input },
    timestamp: new Date(),
  };
}

export function createToolResultMessage(
  id: string,
  toolName: string,
  result?: Record<string, unknown>,
  error?: string
): Message {
  return {
    id,
    kind: "toolResult",
    content: { type: "toolResult", toolName, result, error },
    timestamp: new Date(),
  };
}

export function createSubagentCallMessage(
  id: string,
  agentName: string
): Message {
  return {
    id,
    kind: "subagentCall",
    content: { type: "subagentCall", agentName },
    timestamp: new Date(),
  };
}

export function createSubagentTextMessage(
  id: string,
  agentName: string,
  text: string
): Message {
  return {
    id,
    kind: "subagentText",
    content: { type: "subagentText", agentName, text },
    timestamp: new Date(),
  };
}

export function createSubagentResultMessage(
  id: string,
  agentName: string,
  error?: string
): Message {
  return {
    id,
    kind: "subagentResult",
    content: { type: "subagentResult", agentName, error },
    timestamp: new Date(),
  };
}

export function createGeoFeatureMessage(
  id: string,
  feature: GeoFeature
): Message {
  return {
    id,
    kind: "geoFeature",
    content: { type: "geoFeature", feature },
    timestamp: new Date(),
  };
}
